wagtail/wagtailadmin/views/collections.py

```diff
@@ -1,52 +1,172 @@
+
+        Called for every dispatch event to check if the user has access to the collection view.
+
+        :param wagtail.wagtailcore.models.CollectionPermissionTester collection_perms: The collection permission tester
+            object used to determine the level of access the user should be given.
+        """
+        raise NotImplementedError
+
+    def dispatch(self, request, instance_id, *args, **kwargs):
+        collection = get_object_or_404(Collection, pk=instance_id)
+        collection_perms = collection.permissions_for_user(request.user)
+        if not self.check_permissions(collection_perms):
+            raise PermissionDenied
+        return super(CollectionPermissionMixin, self).dispatch(request, instance_id, *args, **kwargs)
 
 
 class Index(IndexView):
-    permission_policy = collection_permission_policy
     model = Collection
     context_object_name = 'collections'
     template_name = 'wagtailadmin/collections/index.html'
     add_url_name = 'wagtailadmin_collections:add'
-    page_title = ugettext_lazy("Collections")
     add_item_label = ugettext_lazy("Add a collection")
     header_icon = 'folder-open-1'
 
+    def __init__(self):
+        super(Index, self).__init__()
+        self.parent_collection = None
+        self.is_searching = False
+
+    @property
+    def page_title(self):
+        if not self.parent_collection:
+            return ugettext_lazy("Collections")
+        return ugettext_lazy(self.parent_collection.name)
+
     def get_queryset(self):
-        # Only return children of the root node, so that the root is not editable
-        return Collection.get_first_root_node().get_children()
+        search = SearchForm(self.request.GET)
+
+        if search.is_valid() and search.cleaned_data['q']:
+            self.is_searching = True
+            return Collection.objects.filter(name__icontains=search.cleaned_data['q'])
+        else:
+            if not self.parent_collection:
+                # Find the root collection that the user has access to
+                self.parent_collection = get_explorable_root_collection(self.request.user)
+
+            return self.parent_collection.get_children()
+
+    def get_context(self):
+        context = super(Index, self).get_context()
+        # Turn the queryset of collections into a Paginator
+        paginator, collections = paginate(self.request, context[self.context_object_name], per_page=50)
+
+        context.update({
+            'collections': collections,
+            'paginator': paginator,
+            'search_form': SearchForm(),
+        })
+
+        # Chooser search results don't need this
+        if not self.is_searching:
+            context.update({
+                'parent_collection': self.parent_collection,
+                'parent_perms': self.parent_collection.permissions_for_user(self.request.user),
+            })
+
+        return context
+
+    def get(self, request, root_id=None):
```

Comment:
Having this endpoint handle both the browse and (nested) chooser feels a bit complicated, although unfortunately I don't have a cleaner approach off the top of my head. See other comment on `wagtailadmin/views/generic.py`.